### PR TITLE
force field example

### PR DESCRIPTION
**Pull request by ChristopherBiscardi**
# Objective

The Bevy 0.10 release post includes [a great demo](https://bevyengine.org/news/bevy-0-10/#depth-and-normal-prepass) utilizing the depth prepass to power a force field/energy shield type demo. Multiple people have wanted to see the code for this demo over the time since it was first shown, but it has until now only existed in one of @IceSentry 's old branches.

## Solution

Put the demo into the examples folder so that it is not only discoverable, but kept up to date with future Bevy releases.

## Testing

```
cargo run --example force_field
```

---

## Showcase

Primary view

https://github.com/bevyengine/bevy/assets/551247/95f30b0c-839a-4890-9161-3d2bbc8e37aa

Depth prepass view
screenshot-2024-07-08-at-18 50 43@2x

Normal prepass view
screenshot-2024-07-08-at-18 50 47@2x

Motion Vectors prepass view. Only included because we're re-using the prepass shader and it doesn't make sense to *disable* it compared to its usage in the `show_prepass` example.

screenshot-2024-07-08-at-18 50 51@2x

It would actually be nice if there was a "show prepasses" gizmo-ish debug material/plugin thing. 🤔

### CODE REVIEW

File: assets/shaders/force_field.wgsl [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+    // Normalize gradients implicitly by scaling m
+    // Approximation of: m *= inversesqrt( a0*a0 + h*h );
+    m *= 1.79284291400159 - 0.85373472095314 * (a0 * a0 + h * h);
+
+    // Compute final noise value at P
+    let g = vec3(a0.x * x0.x + h.x * x0.y, a0.yz * x12.xz + h.yz * x12.yw);
+    return 130. * dot(m, g);
+}
+
+fn rand2(n: vec2<f32>) -> f32 {
```
Comment (torsteingrindvik) [inline assets/shaders/force_field.wgsl:127]: We have rng functions which are importable: https://github.com/bevyengine/bevy/blob/main/crates/bevy_pbr/src/render/utils.wgsl

File: assets/shaders/force_field.wgsl [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+    x12.x = x12.x - i1.x;
+    x12.y = x12.y - i1.y;
+
+    // Permutations
+    i = mod289(i); // Avoid truncation effects in permutation
+
+    var p = permute3(permute3(i.y + vec3(0., i1.y, 1.)) + i.x + vec3(0., i1.x, 1.));
+    var m = max(0.5 - vec3(dot(x0, x0), dot(x12.xy, x12.xy), dot(x12.zw, x12.zw)), vec3(0.));
+    m *= m;
+    m *= m;
```
Comment (ickshonpe) [inline assets/shaders/force_field.wgsl:109]:
```suggestion
    m = pow(m, vec3(4.));
```

File: examples/shader/force_field.rs [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+    mut commands: Commands,
+    mut meshes: ResMut<Assets<Mesh>>,
+    mut materials: ResMut<Assets<StandardMaterial>>,
+    mut force_field_materials: ResMut<Assets<ForceFieldMaterial>>,
+    mut depth_materials: ResMut<Assets<PrepassOutputMaterial>>,
+) {
+    // plane
+    commands.spawn(PbrBundle {
+        mesh: meshes.add(Rectangle::from_size(Vec2::splat(5.0))),
+        material: materials.add(Color::srgb(0.3, 0.5, 0.3)),
```
Comment (ickshonpe) [inline examples/shader/force_field.rs:45]:
```suggestion
        material: materials.add(Color::from(MEDIUM_SEA_GREEN)),
```

**Review by ickshonpe — COMMENTED** [review]
Gives me a warning when I run it:
```
2024-07-09T12:36:13.101624Z  WARN bevy_pbr::render::mesh_view_bindings: Too many textures in mesh pipeline view layout, this might cause us to hit `wgpu::Limits::max_sampled_textures_per_shader_stage` in some environments.
```
Otherwise seems good. Maybe could have a label showing the current prepass view mode / press space to switch etc.

**Review by alice-i-cecile — APPROVED** [review]

File: examples/shader/force_field.rs [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+            transform: Transform::from_xyz(-0.75, 1.25, 3.0)
+                .looking_at(Vec3::new(2.0, -2.5, -5.0), Vec3::Y),
+            ..default()
+        },
+        NotShadowCaster,
+    ));
+    // light
+    commands.spawn(PointLightBundle {
+        point_light: PointLight {
+            intensity: 1500.0,
```
Comment (rparrett) [inline examples/shader/force_field.rs:85]:
```suggestion
```

This is manually setting a very dim value. (default is 1 million)

**Comment by rparrett** [comment]
Compared to the [original demo](https://bevyengine.org/news/bevy-0-10/#depth-and-normal-prepass), this scene
- seems way too dark (added suggestion)
- has very subtle bloom, which doesn't look quite as impressive (not blocking, I suppose)

I also agree with
> Maybe could have a label showing the current prepass view mode / press space to switch etc.

But I don't mind if someone adds that in a follow up.

**Comment by alice-i-cecile** [comment]
>  I don't personally think this should be an official bevy example. All those features are shown independently in other examples and we already have a ton of examples

- @IceSentry 

I'm relatively convinced by this. I think this is an excellent showcase for a shadertoy sort of tool, but that's not what Bevy's examples are best suited for.

**Comment by ChristopherBiscardi** [comment]
fair enough, thanks for taking a look

**Pull request closed without merging**